### DISCUSSION TITLE
React Folder Structure

### DISCUSSION THREAD

**Discussion by couds** [cw00dw0rd/ArangoBnB#22] [Q&A]
Hi All

I was thinking on. a folder structure like this

```
/src
|-- /componentes (Reusable components across all app like inputs, boxes, accordions, etc)
|-- /screens (like the pages on next but not directly related with routing, more like the type of screen (home, result list, item details, etc)
|-- /services (Here we can create the "connectors" to external services, helpers functions, any non-react oriented code)
|--/store ([OPTIONAL] if we use redux all the redux related code, actions, reducers, constants, ...)
```

And for component level like this

```
component-a/
|-- index.js (reexport the component and apply any HoC)
|-- component-a.js (The code of the component, without any HoC)
|-- styles.scss (I put scss but could by styled or any other)
|-- components/ (This is a personal preference but If a component have smaller components I like to put it inside this folder following the same structure)
```

What do you guys think about this?

Edit: I tried to use the /polls but didn't work T_T, lets use

👍  - Yes I let's use this structure

👎  - No, Let use something else

**Comment by lostpebble** [comment]
Looks good to me. Except I don't like making use of `/index.js` as the export for the component. I prefer to use the full name of the component- as then we avoid "implied" imports in other JS files, like `import ImageComponent from "./components/ImageComponent"`- which is actually importing `/ImageComponent/index.js` instead of what could be confused as `./components/ImageComponent.js`.

So instead, I think we should always use the full component name for the file: `/components/ImageComponent/ImageComponent.js`.

Would rather remove all ambiguity- same goes for default exports. Would prefer if we always name our exports- so not making use of `export default ImageComponent` or whatnot- rather `export const ImageComponent = () => {}`

I see that a lot of people in the React community are finally starting to catch on and work this way now too- it just helps to make things more "static" and direct.

Could be that I work with TypeScript a lot- and these kinds of things just make it easier overall to deal with auto-imports and stuff (helps the IDE know what you actually want when you type it).

**Reply by couds** [reply]
About the index vs component file:

I prefer to use the component kind of like a "black box" so the "implementation" details are hidden, but I can work with that. 

Note: With VSCode the auto import and props autocomplete worked with the index. if you import/export like this

```javascript
import ComponentA from './component-a'

export default ComponentA // here we could also use named export
```

About TS, I do not use it actually, so for me I would like to use JS.

**Reply by lostpebble** [reply]
Yes, I figured we'll use JS as this project should be understood by the most people. You should give TS a go sometime though, its really nice to work with :)

Yea, default exports work fine with most IDEs but they can cause headaches sometimes- I've run into them in the past and I only use named exports in all my projects now.

For example, if I'm in a file like `/page_thing.js` and I try and type:

```js
const Page = () => {
  return (
    <div>
        <ImageCo // <- Typing "ImageComponent" here
    </div>
   );
}
```

Generally, because its a default export the IDE won't know which import you might be referring to, because its not really linked to any real variable- but a file which could be named many different things. I'm sure IDEs do now try and work around this, by looking at the folder / file name- but it feels a bit icky to me. I like to be more deterministic in how things are named and linked.

[This guy](https://humanwhocodes.com/blog/2019/01/stop-using-default-exports-javascript-module/) explains this kinda stuff a lot better.

In any case, since this is just a small project- we can go either way. I can deal with it haha

**Reply by couds** [reply]
I'm thinking on a middle ground. Wdyt inde.js with named export?

**Reply by lostpebble** [reply]
Yea, cool. Lets do that. My dislike for `index.js` isn't as much as default exports 😄

**Reply by lostpebble** [reply]
I just remembered why I don't tend to enjoy `index.js` - because when I use the functionality in my IDE of "Search for filename", and I start searching for "ImageComponent"- it won't find it. And generally these search functions don't search on folder names.

But its all good- not something we should be getting too caught up on anyway.

**Reply by couds** [reply]
Haha 100% agree with the index issue. That's why I suggest to use only for import export. (I had the same issue with the search and hate it.. >.<)

No actual code should be on the index.js file 👍 

I will work this weekend on preparing the folder structure and install the UI library. 

Also try to make a mockups so we can follow for the design (don't expect much on my design skill xD)

**Reply by lostpebble** [reply]
Hahah cool, sounds good!

Should we use something like Figma? Easy to share and collaborate with.

I'm guessing we just focus on Desktop UI ? I know mobile-first is important, but this is just a demo project after all... can get a bit annoying having to balance between the two.

You should also take a look at what @cw00dw0rd has put together on the Vue project already. Spin it up and take a look, he has already got some interface stuff in which might give us a bit of direction.

Let me know once you've got the basic structure in, and I'll add some Store stuff and make the first connection to the backend (API Async action things) using Pullstate for us to work with.
